### DISCUSSION TITLE
How to reproduce Affinity Photo Denoise filter in ImageMagick

### DISCUSSION THREAD

**Discussion by kwiniarski** [ImageMagick/ImageMagick#3889] [Help] [answered]
I'm looking for a way to reproduce Affinity Photo's Denoise filter in ImageMagic, but without success. It seems it works on specific channels of colour space and then "melts" them back with original image. Effect is very good for typical film grain (high ISO images). However, I'm novice to IM and a little bit confused with its options. If you have any advice how to reproduce this as close as possible with IM I would be grateful.

This is how Denoise filter options look like:
![denoise](https://user-images.githubusercontent.com/601868/124019267-223d9280-d9e9-11eb-87fe-dc23ce9ea8a2.png)

And this is how those options affect an image:
![denoise_after_before](https://user-images.githubusercontent.com/601868/124019384-47ca9c00-d9e9-11eb-87b2-e1ae8f13a487.png)

**Comment by fmw42** [comment]
Do you have any details on the algorithm used? Without that, we have little hope of reverse engineering it.

Looks like it might be processing in LAB color space and doing different things on the Luminance and the 2 chrominance AB channels. But what is it doing on each channel type? Is it a wavelet filter?

**Comment by snibgo** [comment]
Affinity Photo is not open source, and I don't use it, so my advice is limited.

Judging from https://affinity.help/photo/en-US.lproj/index.html?page=pages/Filters/filter_denoise.html?title=Denoise, the approach it takes is a common one: separate the image into luminance and chrominance channels, and apply denoising separately.

This approach is useful because luminance (lightness) noise is often more obtrusive than chrominance (colour) noise, and also because the effect of tweaking the sliders can be seen immediately. However, when the source of noise is from a mosaiced camera array, the noise will be propagated by demosaicing, making it more difficult to fix. In those cases, denoising before demosaicing is more effective. For interactive editors, the image needs to be demosaiced whenever a denoise slider is pushed, which would be a major performance issue, so interactive editors don't usually work this way.

Anyhow, whether we denoise RGB channels or Lab or JzAzBz channels, many methods are available to denoise each channel: wavelet, guided filter, selective blur, limit min-max, and variations of blurring. Any of these may be used with a mask. Ideally, the method used depends on the cause and nature of the noise. See, for example, [Camera noise](http://im.snibgo.com/camnoise.htm), which uses photon shot noise from a DSLR as an example.

Without knowing more about Affinity, I can't advise on the closest method, nor on exactly what the Affinity sliders do.

**Comment by kwiniarski** [comment]
Thank you for your replays. Well, I might not express myself correctly asking wrong question. I don't expect to do reverse engineering on Affinity. But as I guess, and what @snibgo suggests, their filter works on specific colour channels. In fact, other editors seem do the same. For example, open-source RawTherapee has this denoise filter: http://rawpedia.rawtherapee.com/Noise_Reduction. It seems to do the same thing, and on my test image it looks even better then Affinity. Of course, in Photoshop you can do this as well: https://www.photoshopessentials.com/photo-editing/reduce-noise/ or https://www.exposureguide.com/image-noise-reduction/.
In the end I was looking for a way to do something similar in IM.

So maybe I should start from basics? How to apply some filters, for example simple blur, to specific color channel and apply it to original image? Do I have to work with some intermediate images (but how to correctly apply them back to the original image) or maybe I can do this with some layers (`-clone` and `-composite`)?

Sorry if those are silly questions, but I couldn't find any decent tutorial on this and so far, I was using IM for simple operations like batch rescale or auto-levels, etc.

**Comment by snibgo** [comment]
So, you are not a total beginner with IM.

I have linked to [Camera noise](http://im.snibgo.com/camnoise.htm), which shows methods for denoising channels with IM.

I suggest you start with very simple commands, such as:

1. Read an image, blur the green channel, and write it.

2. Read an image, convert to Lab colorspace, convert back to sRGB, and write it.

3. As 2, but blur the L channel.

4. As 2, but apply the "-wavelet-denoise" operation to the L channel. See https://imagemagick.org/script/command-line-options.php#wavelet-denoise

**Reply by kwiniarski** [reply]
Thank you for the suggestions and link to the article. I have looked over different scripts from the internet and recombined them into something like this:
```cmd
magick.exe in.jpg ^
( -clone 0 -colorspace lab -channel 0 -separate ) ^
( -clone 0 -wavelet-denoise 4096x0 -colorspace lab -separate -delete 0 ) ^
-delete 0 -set colorspace lab -combine ^
-colorspace srgb ^
( -clone 0 -despeckle ) ^
( -clone 0 -wavelet-denoise 4096x0 ) ^
-evaluate-sequence median ^
-adaptive-sharpen 0x1 ^
out.jpg
```
Result is satisfactory:
![compare](https://user-images.githubusercontent.com/601868/124347878-17505100-dbe7-11eb-891c-83800b546dac.jpg)

However, I have the feeling that above script is not optimal. Mainly second part after `-colorspace srgb`. Operations on Lab color space remove many colors related grain, but rest is left:
![compare-partial](https://user-images.githubusercontent.com/601868/124348167-b1fd5f80-dbe8-11eb-9572-4d178c550a92.jpg)

I couldn't get rid of it with just `-selective-blur` or `-wavelet-denoise` as too much detail was lost. Also editing other channels didn't help. Then I have found some articles about removing noise with stacked images. Normally I would need the same photo shoot in series with different ISO. But I have only one, so I cloned original with different denoising methods to average pixels, and then flatten them using median pixel value. To do this I needed at least two more images, to get median from three. And it did the job! But is it optimal? Maybe there is a better way to get rid of this grain that has left after first part?

**Comment by snibgo** [comment, marked as answer]
The effect of the first few lines is to wavelet-denoise the a\* and b\* channels, with no change to the L\* channel. As noise in the L\* channel is the most obtrusive, I don't know why you do this.

Then you take the median of three images. The first is a denoised version, the second is a despeckled version of the first, and the third is a wavelet-denoised version of the first. Despeckling has a weaker denoising effect than your wavelet-denoised operation. So the median is almost exactly the same as the depeckled version. So creating the three versions and finding their median is almost totally wasted effort. You could simply despeckle instead.

You might consider wavelet-denoising the L\*a\*b\* image, and blending the denoised image with the input, using a different blend for each channel. (But the blend factors for the a\* and b\* channels should usually be the same.) Windows BAT script:
```
magick ^
  %1 ^
  -alpha off ^
  -colorspace Lab ^
  -separate ^
  ( -clone 0,1,2 ^
    -wavelet-denoise 4096x0 ^
  ) ^
  ( -clone 0,3 ^
    -compose Blend -define compose:args=%3 -composite ) ^
  ( -clone 1,4 ^
    -compose Blend -define compose:args=%4 -composite ) ^
  ( -clone 2,5 ^
    -compose Blend -define compose:args=%5 -composite ) ^
  -delete 0-5 ^
  -set colorspace Lab -combine ^
  -colorspace sRGB ^
  %2
```

The blend factors should usually be in the range 0 to 100. You can run that script to explore the effect of denoising just L\*, or just a\* and b\*, or all three.

And you can play with the denoise parameters "4096x0".

I suggest you do that without the unsharp at the end, because that will confuse comparisons. When it does what you want, you can add an unsharp if you want.

**Reply by kwiniarski** [reply]
Thank you for the script. It works perfectly! I have only modified first clone operation, by cloning L separately and a with b separately. L seems to be very sensitive to denoise and low values do not give any change and too high create blurred image. Interestingly changing blend amount doesn't help much. Seems like for channel 0 reasonable wavelet threshold is around 3-5%, and blend around 30-50. Form the other hand, other two channels play well with very high values. Values bellow 50% for those channels, for both threshold and blend, seem to disable effects of this operation completely - they are unnoticeable. At least on my test images. Thank you once again.
